### PR TITLE
[BEAM-2963] Pass portable transform IDs in Dataflow translation

### PR DESCRIPTION
**Pull request by kennknowles**
This changes the Dataflow translation to reference the portable pipeline by id rather than putting serialized Java bits directly in the Dataflow job.

----

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

**Comment by lukecwik** [comment]
Reviewed 17 of 17 files at r1.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ParDoTranslation.java, line 140 at r1](https://reviewable.io:443/reviews/apache/beam/4356#-L2MYqWIr4ZW5LM1S2q3:-L2MYqWIr4ZW5LM1S2q4:bdazttx) ([raw file](https://github.com/apache/beam/blob/5a38bdee4d640967ccbcfaa82516e448d3b1ef66/runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ParDoTranslation.java#L140)):*
> ```Java
>   }
> 
>   public static ParDoPayload translateParDo(
> ```

You changed all the names of the `to` methods. Were you planning on changing the `from` method names as well?
Doesn't have to be this PR, just wondering why you wanted to rename all the methods.

---

*[runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ParDoTranslation.java, line 150 at r1](https://reviewable.io:443/reviews/apache/beam/4356#-L2MWFI_eAcqwtP4bUiY:-L2MWFI_eAcqwtP4bUiZ:bt3vki6) ([raw file](https://github.com/apache/beam/blob/5a38bdee4d640967ccbcfaa82516e448d3b1ef66/runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ParDoTranslation.java#L150)):*
> ```Java
>           @Override
>           public SdkFunctionSpec translateDoFn(SdkComponents newComponents) {
>             return ParDoTranslation.translateDoFn(parDo.getFn(), parDo.getMainOutputTag());
> ```

Drop the `ParDoTranslation` qualifier.

---

*[runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ParDoTranslation.java, line 456 at r1](https://reviewable.io:443/reviews/apache/beam/4356#-L2MZw_z1wvCVHSVBdEm:-L2MZw_z1wvCVHSVBdEn:b7cnvsp) ([raw file](https://github.com/apache/beam/blob/5a38bdee4d640967ccbcfaa82516e448d3b1ef66/runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ParDoTranslation.java#L456)):*
> ```Java
>   }
> 
>   // Using nullability instead of optional because of shading
> ```

Use method comment explaining the purpose of null return value.

---

*[runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java, line 170 at r1](https://reviewable.io:443/reviews/apache/beam/4356#-L2MfmAFiLwV0d91-3Jx:-L2MfmAFiLwV0d91-3Jy:b125ob2) ([raw file](https://github.com/apache/beam/blob/5a38bdee4d640967ccbcfaa82516e448d3b1ef66/runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java#L170)):*
> ```Java
>     RunnerApi.Pipeline pipelineProto = PipelineTranslation.toProto(pipeline, sdkComponents);
> 
>     LOG.info("Portable pipeline proto:\n{}", TextFormat.printToString(pipelineProto));
> ```

Debug log level?

This will spam all our current users otherwise.

---

*[runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java, line 521 at r1](https://reviewable.io:443/reviews/apache/beam/4356#-L2Mc-6nOr4gZQZwC1bG:-L2Mc-6nOr4gZQZwC1bH:b-kvgsmq) ([raw file](https://github.com/apache/beam/blob/5a38bdee4d640967ccbcfaa82516e448d3b1ef66/runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java#L521)):*
> ```Java
> 
>     @Override
>     public SdkComponents getSdkComponents() {
> ```

Why does this need to be make public?

---

*[runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java, line 685 at r1](https://reviewable.io:443/reviews/apache/beam/4356#-L2McWJHhpyh3RGIPW7d:-L2McWJHhpyh3RGIPW7e:b3u41zk) ([raw file](https://github.com/apache/beam/blob/5a38bdee4d640967ccbcfaa82516e448d3b1ef66/runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java#L685)):*
> ```Java
>     // Stage the pipeline, retrieving the staged pipeline path, then update
>     // the options on the new job
>     LOG.info("Staging pipeline description to {}", options.getStagingLocation());
> ```

Label this with a TODO to make Dataflow accept an explicit "pipeline" parameter. In the long term we should migrate Dataflow to use the Apache Beam pipeline representation.

---

*[runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/PrimitiveParDoSingleFactory.java, line 113 at r1](https://reviewable.io:443/reviews/apache/beam/4356#-L2Mii3LDuSxAByA4WLa:-L2Mii3LDuSxAByA4WLb:b32j807) ([raw file](https://github.com/apache/beam/blob/5a38bdee4d640967ccbcfaa82516e448d3b1ef66/runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/PrimitiveParDoSingleFactory.java#L113)):*
> ```Java
>   }
> 
>   /** A translator for {@link ParDoSingle} */
> ```

Can we not use `ParDoSingleViaMulti` and use the existing translation inside ParDoTranslation?

---

*[runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java, line 166 at r1](https://reviewable.io:443/reviews/apache/beam/4356#-L2MgsWdYzTYzvaBdFqM:-L2MgsWdYzTYzvaBdFqN:b31qk6t) ([raw file](https://github.com/apache/beam/blob/5a38bdee4d640967ccbcfaa82516e448d3b1ef66/runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java#L166)):*
> ```Java
>     return p;
>   }
> ```

This file did not change in any meaningful way other then whitespace. If you want to fix formatting, a separate PR would have made more sense.

---

*[sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ParDo.java, line 584 at r1](https://reviewable.io:443/reviews/apache/beam/4356#-L2MgHGZDCUFH7KUW09j:-L2MgHGZDCUFH7KUW09k:bntwqi5) ([raw file](https://github.com/apache/beam/blob/5a38bdee4d640967ccbcfaa82516e448d3b1ef66/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ParDo.java#L584)):*
> ```Java
>       extends PTransform<PCollection<? extends InputT>, PCollection<OutputT>> {
> 
>     public static final String MAIN_OUTPUT_TAG = "output";
> ```

I don't think we want to make this public. Seems pretty easy to just copy the constant or place it in pipeline translation instead of polluting a users auto complete.

---

*[sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java, line 990 at r1](https://reviewable.io:443/reviews/apache/beam/4356#-L2MeiEWQPAqC3i2UhI3:-L2MeiEWQPAqC3i2UhI4:by32h39) ([raw file](https://github.com/apache/beam/blob/192b4c70927901860312f8c8acd27bd47e4a4259/sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java#L990)):*
> ```Java
>       ByteString.Output encodedKeyOut = ByteString.newOutput();
> 
>       Coder<K> keyCoder = ((KvCoder<K, ?>) inputCoder.getValueCoder()).getKeyCoder();
> ```

Please update `createOrUseCachedBagUserStateKey` to use `getValueCoder` instead of `getCoderArguments`

---

*[sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java, line 189 at r1](https://reviewable.io:443/reviews/apache/beam/4356#-L2MfRg7d6PTNTGKz0FY:-L2MfL2feA2REWV4Glkp:b-1xmwrh) ([raw file](https://github.com/apache/beam/blob/5a38bdee4d640967ccbcfaa82516e448d3b1ef66/sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java#L189)):*
> ```Java
>     }
> 
>     String pTransformString = TextFormat.printToString(pTransform);
> ```

Please only do this conversion lazily.

---

*[sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java, line 196 at r1](https://reviewable.io:443/reviews/apache/beam/4356#-L2Mf94NSiKqKPhASHE0:-L2Mf94NSiKqKPhASHE1:b8bqqdz) ([raw file](https://github.com/apache/beam/blob/5a38bdee4d640967ccbcfaa82516e448d3b1ef66/sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java#L196)):*
> ```Java
> 
>     checkArgument(pTransform.getSubtransformsCount() == 0,
>       "Cannot process composite transform: %s",
> ```

nit: whitespace

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/apache/beam/4356)*

**Comment by kennknowles** [comment]
Looks like the same breakage from hcatalog on all my PRs. Known?

**Comment by kennknowles** [comment]
It also looks like findbugs under gradle puts output not in the Jenkins log but a file and logs a local filesystem URL.

**Comment by lukecwik** [comment]
hcatalog was one of the hardest modules I attempted to shade because of how many dependencies bundled without relocating things.

**Comment by kennknowles** [comment]
Reviewed 17 of 17 files at r1.
Review status: 16 of 17 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ParDoTranslation.java, line 140 at r1](https://reviewable.io:443/reviews/apache/beam/4356#-L2MYqWIr4ZW5LM1S2q3:-L2XHjNVUKsrN3c9_JnZ:b-lvb113) ([raw file](https://github.com/apache/beam/blob/5a38bdee4d640967ccbcfaa82516e448d3b1ef66/runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ParDoTranslation.java#L140)):*
<details><summary><i>Previously, lukecwik (Lukasz Cwik) wrote…</i></summary><blockquote>

You changed all the names of the `to` methods. Were you planning on changing the `from` method names as well?
Doesn't have to be this PR, just wondering why you wanted to rename all the methods.
</blockquote></details>

Yea, the readability of them all having the same name was getting pretty bad. I don't have a great idea of what to rename the other ones but I would definitely like to do so.

---

*[runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ParDoTranslation.java, line 150 at r1](https://reviewable.io:443/reviews/apache/beam/4356#-L2MWFI_eAcqwtP4bUiY:-L2XITM7R64uycAS30uQ:bgefpqq) ([raw file](https://github.com/apache/beam/blob/5a38bdee4d640967ccbcfaa82516e448d3b1ef66/runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ParDoTranslation.java#L150)):*
<details><summary><i>Previously, lukecwik (Lukasz Cwik) wrote…</i></summary><blockquote>

Drop the `ParDoTranslation` qualifier.
</blockquote></details>

I can't, because this method itself is called `translateDoFn`, from a different context.

---

*[runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ParDoTranslation.java, line 456 at r1](https://reviewable.io:443/reviews/apache/beam/4356#-L2MZw_z1wvCVHSVBdEm:-L2XIsjHpcOB_PV2EZG8:b-896fix) ([raw file](https://github.com/apache/beam/blob/5a38bdee4d640967ccbcfaa82516e448d3b1ef66/runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ParDoTranslation.java#L456)):*
<details><summary><i>Previously, lukecwik (Lukasz Cwik) wrote…</i></summary><blockquote>

Use method comment explaining the purpose of null return value.
</blockquote></details>

Done.

---

*[runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java, line 170 at r1](https://reviewable.io:443/reviews/apache/beam/4356#-L2MfmAFiLwV0d91-3Jx:-L2XJpT4Ps48sqHU14lX:b8lwn5d) ([raw file](https://github.com/apache/beam/blob/5a38bdee4d640967ccbcfaa82516e448d3b1ef66/runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java#L170)):*
<details><summary><i>Previously, lukecwik (Lukasz Cwik) wrote…</i></summary><blockquote>

Debug log level?

This will spam all our current users otherwise.
</blockquote></details>

This will not spam. It is once per job launch. It is at least as important as the fusion messages we currently print.

---

*[runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java, line 521 at r1](https://reviewable.io:443/reviews/apache/beam/4356#-L2Mc-6nOr4gZQZwC1bG:-L2XKPdixjjppDmEuxVy:bva4au2) ([raw file](https://github.com/apache/beam/blob/5a38bdee4d640967ccbcfaa82516e448d3b1ef66/runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java#L521)):*
<details><summary><i>Previously, lukecwik (Lukasz Cwik) wrote…</i></summary><blockquote>

Why does this need to be make public?
</blockquote></details>

The interface has it public; cannot be narrowed here.

---

*[runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java, line 685 at r1](https://reviewable.io:443/reviews/apache/beam/4356#-L2McWJHhpyh3RGIPW7d:-L2XKtWqTvo72307BnzN:b-896fix) ([raw file](https://github.com/apache/beam/blob/5a38bdee4d640967ccbcfaa82516e448d3b1ef66/runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java#L685)):*
<details><summary><i>Previously, lukecwik (Lukasz Cwik) wrote…</i></summary><blockquote>

Label this with a TODO to make Dataflow accept an explicit "pipeline" parameter. In the long term we should migrate Dataflow to use the Apache Beam pipeline representation.
</blockquote></details>

Done.

---

*[runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/PrimitiveParDoSingleFactory.java, line 113 at r1](https://reviewable.io:443/reviews/apache/beam/4356#-L2Mii3LDuSxAByA4WLa:-L2XKyLhQ9t6uJKbEHMO:b-363dwj) ([raw file](https://github.com/apache/beam/blob/5a38bdee4d640967ccbcfaa82516e448d3b1ef66/runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/PrimitiveParDoSingleFactory.java#L113)):*
<details><summary><i>Previously, lukecwik (Lukasz Cwik) wrote…</i></summary><blockquote>

Can we not use `ParDoSingleViaMulti` and use the existing translation inside ParDoTranslation?
</blockquote></details>

Because of Dataflow's limitation and UI goals, we override SingleViaMulti to be a Dataflow-specific Single. That way it can have display data (composites cannot have display data) and also shows up non-nested in the UI.

---

*[runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java, line 166 at r1](https://reviewable.io:443/reviews/apache/beam/4356#-L2MgsWdYzTYzvaBdFqM:-L2XLNMh82vw0qiZN4FM:b-crnxyy) ([raw file](https://github.com/apache/beam/blob/5a38bdee4d640967ccbcfaa82516e448d3b1ef66/runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java#L166)):*
<details><summary><i>Previously, lukecwik (Lukasz Cwik) wrote…</i></summary><blockquote>

This file did not change in any meaningful way other then whitespace. If you want to fix formatting, a separate PR would have made more sense.
</blockquote></details>

Reverted. It changed (addition of a parameter) and then changed back (removal of a parameter) so I had autoformatted it.

---

*[sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ParDo.java, line 584 at r1](https://reviewable.io:443/reviews/apache/beam/4356#-L2MgHGZDCUFH7KUW09j:-L2XLx2e-KZPUX9UrHiT:by3w2c6) ([raw file](https://github.com/apache/beam/blob/5a38bdee4d640967ccbcfaa82516e448d3b1ef66/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ParDo.java#L584)):*
<details><summary><i>Previously, lukecwik (Lukasz Cwik) wrote…</i></summary><blockquote>

I don't think we want to make this public. Seems pretty easy to just copy the constant or place it in pipeline translation instead of polluting a users auto complete.
</blockquote></details>

Good catch - it should be `@Internal` at least. I wanted to avoid a magic string, hence using a shared reference. But this can be done in a way that is more encapsulated in the DataflowRunner. Done, see `PrimitiveParDoSingleFactory`. I actually like keeping it here so that the tag is readable in debugging so I did leave it private.

---

*[sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java, line 990 at r1](https://reviewable.io:443/reviews/apache/beam/4356#-L2MeiEWQPAqC3i2UhI3:-L2XP4uIGltmzn5JmoZx:bxncrr6) ([raw file](https://github.com/apache/beam/blob/192b4c70927901860312f8c8acd27bd47e4a4259/sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java#L990)):*
<details><summary><i>Previously, lukecwik (Lukasz Cwik) wrote…</i></summary><blockquote>

Please update `createOrUseCachedBagUserStateKey` to use `getValueCoder` instead of `getCoderArguments`
</blockquote></details>

Done, if I understood what you mean.

---

*[sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java, line 189 at r1](https://reviewable.io:443/reviews/apache/beam/4356#-L2MfRg7d6PTNTGKz0FY:-L2XQQQl_J7GDupG3m28:b-vc12d3) ([raw file](https://github.com/apache/beam/blob/5a38bdee4d640967ccbcfaa82516e448d3b1ef66/sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java#L189)):*
<details><summary><i>Previously, lukecwik (Lukasz Cwik) wrote…</i></summary><blockquote>

Please only do this conversion lazily.
</blockquote></details>

Done. TBH I don't know how to `checkArgument(pTransform.hasSpec(), "%s", () -> TextFormat.printToString(pTransform))`

But this is once per pipeline execution, more-or-less, right?

---

*[sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java, line 196 at r1](https://reviewable.io:443/reviews/apache/beam/4356#-L2Mf94NSiKqKPhASHE0:-L2XRSNeQgl8uj_SYngi:b-896fix) ([raw file](https://github.com/apache/beam/blob/5a38bdee4d640967ccbcfaa82516e448d3b1ef66/sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java#L196)):*
<details><summary><i>Previously, lukecwik (Lukasz Cwik) wrote…</i></summary><blockquote>

nit: whitespace
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/apache/beam/4356)*

**Comment by kennknowles** [comment]
Not strictly part of this change, but included a fix for `getAdditionalInputs` of the primitive single-output override.

**Comment by lukecwik** [comment]
Review status: 10 of 16 files reviewed at latest revision, 9 unresolved discussions.

---

*[runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java, line 170 at r1](https://reviewable.io:443/reviews/apache/beam/4356#-L2MfmAFiLwV0d91-3Jx:-L2gCaVdVU4Matc-G2ja:b4riwhc) ([raw file](https://github.com/apache/beam/blob/5a38bdee4d640967ccbcfaa82516e448d3b1ef66/runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java#L170)):*
<details><summary><i>Previously, kennknowles (Kenn Knowles) wrote…</i></summary><blockquote>

This will not spam. It is once per job launch. It is at least as important as the fusion messages we currently print.
</blockquote></details>

I tried this out and this 100's if not 1000's of lines (depending on pipeline complexity). Please swap to debug.

---

*[sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java, line 189 at r1](https://reviewable.io:443/reviews/apache/beam/4356#-L2MfRg7d6PTNTGKz0FY:-L2gE7d8IBBzekWEoEon:bfatf5b) ([raw file](https://github.com/apache/beam/blob/5a38bdee4d640967ccbcfaa82516e448d3b1ef66/sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java#L189)):*
<details><summary><i>Previously, kennknowles (Kenn Knowles) wrote…</i></summary><blockquote>

Done. TBH I don't know how to `checkArgument(pTransform.hasSpec(), "%s", () -> TextFormat.printToString(pTransform))`

But this is once per pipeline execution, more-or-less, right?
</blockquote></details>

No, the process bundle handler is for every process bundle request unless the ProcessBundleHandler is re-used across bundles.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/apache/beam/4356)*

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java [hunk 1]
```diff
@@ -157,22 +159,25 @@ private Pipeline buildPipeline(DataflowPipelineOptions options) {
     FileSystems.setDefaultPipelineOptions(options);
 
     p.apply("ReadMyFile", TextIO.read().from("gs://bucket/object"))
-        .apply("WriteMyFile", TextIO.write().to("gs://bucket/object"));
+     .apply("WriteMyFile", TextIO.write().to("gs://bucket/object"));
```
Comment (lukecwik) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java]: nit: whitespace

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java [hunk 2]
```diff
@@ -1381,4 +1381,51 @@ public void checkNotNull(Object value, String message, Object... args) {
       }
     }
   }
+
+  public static StateSpec<?> getStateSpecOrCrash(
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java]: getStateSpecOrThrow?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java [hunk 3]
```diff
@@ -1381,4 +1381,51 @@ public void checkNotNull(Object value, String message, Object... args) {
...
+      throw new RuntimeException(
+          String.format(
+              "Malformed %s class %s: state declaration field %s is not accessible.",
+              DoFn.class.getSimpleName(),
+              target.getClass().getName(),
+              stateDeclaration.field().getName()));
+    }
+  }
+
+  public static TimerSpec getTimerSpecOrCrash(
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java]: getTimerSpecOrThrow

**Comment by kennknowles** [comment]
Review status: 10 of 16 files reviewed at latest revision, 9 unresolved discussions.

---

*[runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java, line 170 at r1](https://reviewable.io:443/reviews/apache/beam/4356#-L2MfmAFiLwV0d91-3Jx:-L2gIySoLwK1HiY4NnOQ:b-k8it61) ([raw file](https://github.com/apache/beam/blob/5a38bdee4d640967ccbcfaa82516e448d3b1ef66/runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java#L170)):*
<details><summary><i>Previously, lukecwik (Lukasz Cwik) wrote…</i></summary><blockquote>

I tried this out and this 100's if not 1000's of lines (depending on pipeline complexity). Please swap to debug.
</blockquote></details>

Fair enough. I shouldn't use the fact that our current output is a spam of debug-only info as an excuse to add more. To have a decent CLI to expose this, we should support `--verbose` or `--debug` etc, so a user (or contributor) doesn't have to tweak a logging backend, which almost universally have terrible interfaces for one-off adjustments. Filed https://issues.apache.org/jira/browse/BEAM-3475.

---

*[sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java, line 189 at r1](https://reviewable.io:443/reviews/apache/beam/4356#-L2MfRg7d6PTNTGKz0FY:-L2gIgAz41sESHK5sw6g:btao1bh) ([raw file](https://github.com/apache/beam/blob/5a38bdee4d640967ccbcfaa82516e448d3b1ef66/sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java#L189)):*
<details><summary><i>Previously, lukecwik (Lukasz Cwik) wrote…</i></summary><blockquote>

No, the process bundle handler is for every process bundle request unless the ProcessBundleHandler is re-used across bundles.
</blockquote></details>

That seems problematic in and of itself. This logic should all happen during registration, since it is basic one-time validation of the process bundle descriptor. But it can be fixed later if/when it shows up in performance info, I guess.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/apache/beam/4356)*

**Comment by kennknowles** [comment]
Reviewed 6 of 7 files at r3, 1 of 1 files at r4.
Review status: 15 of 16 files reviewed at latest revision, 5 unresolved discussions.

---

*[sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java, line 1385 at r4](https://reviewable.io:443/reviews/apache/beam/4356#-L2MW2Ot6QEYmR9VRXfu-r4-1385:-L2gLo8_8DB57wersnoR:b-896fix) ([raw file](https://github.com/apache/beam/blob/da968143f5c16c5e5b3ebc31e37d252bc97b5b4a/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java#L1385)):*
<details><summary><i>Previously, lukecwik (Lukasz Cwik) wrote…</i></summary><blockquote>

getStateSpecOrThrow?
</blockquote></details>

Done.

---

*[sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java, line 1408 at r4](https://reviewable.io:443/reviews/apache/beam/4356#-L2MW2Ot6QEYmR9VRXfu-r4-1408:-L2gLoUXfUPMY2bIjZkm:b-896fix) ([raw file](https://github.com/apache/beam/blob/da968143f5c16c5e5b3ebc31e37d252bc97b5b4a/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java#L1408)):*
<details><summary><i>Previously, lukecwik (Lukasz Cwik) wrote…</i></summary><blockquote>

getTimerSpecOrThrow
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/apache/beam/4356)*

**Comment by kennknowles** [comment]
Thanks! I'll squash and let tests run to make sure I didn't mess anything up while commuting commits.

**Comment by lukecwik** [comment]
Review status: 12 of 16 files reviewed at latest revision, 5 unresolved discussions.

---

*[sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java, line 189 at r1](https://reviewable.io:443/reviews/apache/beam/4356#-L2MfRg7d6PTNTGKz0FY:-L2gPzOm3JTqeBcwZDI5:bgvgh4c) ([raw file](https://github.com/apache/beam/blob/5a38bdee4d640967ccbcfaa82516e448d3b1ef66/sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java#L189)):*
<details><summary><i>Previously, kennknowles (Kenn Knowles) wrote…</i></summary><blockquote>

That seems problematic in and of itself. This logic should all happen during registration, since it is basic one-time validation of the process bundle descriptor. But it can be fixed later if/when it shows up in performance info, I guess.
</blockquote></details>

You can move the validation logic within the register handler then which is only invoked once per registration request.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/apache/beam/4356)*

**Comment by kennknowles** [comment]
Thanks! I'll squash and let tests run to make sure I didn't mess anything up while commuting commits.

**Comment by kennknowles** [comment]
Rebased on hopefully-now-green `master`. I will merge when I get test confirmation now.

**Comment by kennknowles** [comment]
Run Dataflow ValidatesRunner

**Comment by kennknowles** [comment]
@tgroh I think this will resolve some shading pain you are encountering, too.

**Comment by kennknowles** [comment]
Run Dataflow ValidatesRunner

**Pull request merged**